torchmultimodal/models/video_vqvae.py

```diff
@@ -4,10 +4,162 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
+from typing import Any, Dict, List, Tuple, Union
+
 from torch import nn, Tensor
+from torch.nn import functional as F
 from torchmultimodal.modules.layers.attention import AxialAttentionBlock
 
-from torchmultimodal.modules.layers.conv import SamePadConv3d
+from torchmultimodal.modules.layers.conv import SamePadConv3d, SamePadConvTranspose3d
+
+
+class VideoEncoder(nn.Module):
+    """Encoder for Video VQVAE. Stacks specified number of ``SamePadConv3d`` layers
+    followed by a stack of ``AttentionResidualBlocks`` and a final ``SamePadConv3d``
+    layer before the codebook. The residual blocks use Axial Attention to enhance
+    representations of video data without significantly increasing computational
+    cost. Follows VideoGPT's implementation:
+    https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/vqvae.py
+
+    Args:
+        in_channels (List[int]): list of input channel dimension for each conv layer
+        out_channels (List[int]): list of output channel dimension for each conv layer
+        kernel_sizes (List[int or Tuple[int]]): list of kernel sizes for each conv layer
+        strides (List[int or Tuple[int]]): list of strides for each conv layer
+        n_res_layers (int): number of ``AttentionResidualBlocks`` to include
+        embedding_dim (int): size of hidden dimension of final output
+        **kwargs (dict): keyword arguments to be passed into ``SamePadConv3d`` and used by ``nn.Conv3d``
+
+    Raises:
+        ValueError: if the lengths of ``in_channels``, ``out_channels``, ``kernel_sizes``,
+                    and ``strides`` are not all equivalent
+        ValueError: if ``in_channels`` is not identical to ``out_channels`` offset by one
+
+    Inputs:
+        x (Tensor): input video data with shape (b x c x d1 x d2 x d3)
+    """
+
+    def __init__(
+        self,
+        in_channels: List[int],
+        out_channels: List[int],
+        kernel_sizes: List[Union[int, Tuple[int, int, int]]],
+        strides: List[Union[int, Tuple[int, int, int]]],
+        n_res_layers: int,
+        embedding_dim: int,
+        **kwargs: Dict[str, Any],
+    ):
+        super().__init__()
+        if not (
+            len(in_channels) == len(out_channels) == len(kernel_sizes) == len(strides)
+        ):
+            raise ValueError(
+                "in_channels, out_channels, kernel_sizes, strides should all have the same length"
+            )
+        if in_channels[1:] != out_channels[:-1]:
+            raise ValueError("out_channels should match in_channels offset by one")
```

Comment:
I find we are doing these checks a lot for conv layers. Maybe we should write a util for it; I guess the tradeoff is that it may come at the cost of code readability. Wdyt?

Reply:
yup agree, was thinking of doing this in a separate PR